test/test_asarray.py

```diff
@@ -0,0 +1,423 @@
+    def test_copy_from_dlpack(self, device, dtype):
+        if dtype is torch.bool:
+            return
+
+        original = common.make_tensor((5, 5), device, dtype)
+
+        def check(**kwargs):
+            self._check(original, to_dlpack, is_alias=False, **kwargs)
+
+        check(requires_grad=self._may_compute_grad(dtype), copy=True)
+        check(dtype=dtype, copy=True)
+        check(device=device, copy=True)
+        check(device=device, dtype=dtype, copy=True)
+
+        # Copy is forced because of different device
+        if torch.cuda.is_available():
+            other = "cuda" if device == "cpu" else "cpu"
+            check(same_device=False, device=other)
+            check(same_device=False, device=other, copy=True)
+            check(same_device=False, device=other, dtype=dtype, copy=True)
+
+        # Copy is forced because of different dtype
+        for other in get_all_dtypes():
+            if dtype != other:
+                check(same_dtype=False, dtype=other)
+                check(same_dtype=False, dtype=other, copy=True)
+
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_alias_from_buffer(self, device, dtype):
+        original = common.make_tensor((5,), device, dtype).numpy()
+
+        def check(**kwargs):
+            self._check(original, **kwargs)
+
+        check(dtype=dtype, requires_grad=self._may_compute_grad(dtype))
+        check(dtype=dtype, copy=False)
+        check(device=device, dtype=dtype)
+        check(device=device, dtype=dtype, copy=False)
+
+    @onlyCPU
+    @dtypes(*common.torch_to_numpy_dtype_dict.keys())
+    def test_copy_from_buffer(self, device, dtype):
+        original = common.make_tensor((5,), device, dtype).numpy()
+
+        def check(**kwargs):
+            self._check(original, is_alias=False, **kwargs)
+
+        check(dtype=dtype, requires_grad=self._may_compute_grad(dtype), copy=True)
+        check(dtype=dtype, device=device, copy=True)
+
+        # Copy is forced because of different device
+        if torch.cuda.is_available():
+            check(same_device=False, dtype=dtype, device="cuda")
+            check(same_device=False, dtype=dtype, device="cuda", copy=True)
+
+    @dtypes(*get_all_dtypes())
+    def test_copy_list(self, device, dtype):
+        original = common.make_tensor((5, 5), device, dtype)
+
+        def check(**kwargs):
+            self._check(original, torch.Tensor.tolist, is_alias=False, **kwargs)
+
+        check(device=device, dtype=dtype)
+        check(device=device, dtype=dtype, requires_grad=self._may_compute_grad(dtype))
+        check(device=device, dtype=dtype, copy=True)
+
+    @dtypes(*get_all_dtypes())
+    def test_unsupported_alias(self, device, dtype):
+        original = common.make_tensor((5, 5), device, dtype)
+
+        if torch.cuda.is_available():
+            other_device = "cuda" if device == "cpu" else "cpu"
+
+            with self.assertRaisesRegex(ValueError,
```

Comment:
Previously I suggested asserts like this appear earlier, but it's OK to collect them here, too